cdap-data-fabric/src/main/java/co/cask/cdap/spi/data/StructuredTable.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.spi.data;
+
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.spi.data.table.StructuredTableSpecification;
+import co.cask.cdap.spi.data.table.field.Field;
+import co.cask.cdap.spi.data.table.field.Range;
+
+import java.util.Collection;
+import java.util.Optional;
+
+/**
+ * Abstraction for a table that contains rows and columns.
+ * The schema of the table is fixed, and has to be specified in the
+ * {@link StructuredTableSpecification} during the table creation.
+ */
+// TODO: Add IOException for table operations if needed when implementing SQL tables.
+public interface StructuredTable {
+  /**
+   * Write the collection of fields to the table.
+   * The fields contain both the primary key and the rest of the columns to write.
+   *
+   * @param fields the fields to write
+   * @throws InvalidFieldException if any of the fields are not part of the table schema, or the types do not match.
+   */
+  void write(Collection<Field<?>> fields) throws InvalidFieldException;
+
+  /**
+   * Read a single row with all the columns from the table.
+   *
+   * @param keys the primary key of the row to read
+   * @return the row addressed by the primary key
```

Comment:
Not exactly accurate about the return type. It actually an optional that if it is not empty, then it contains the row addressed by the key; otherwise the given key is missing in the table.

Reply:
Clarified